Add support for Union type in structured configs.

pereman2 opened the pull request:
To add support for UnionTypes a new type of node is created `UnionNode`. This node holds a value which is another node so it's easier to check values. I plan to modify `_validate_value` which validates that the input is correct with the values implementation of `validate_and_convert` because UnionNode should not know what a ListConfig or DictConfig is. 

I added some tests to check the behaviour. At this time I don't know which ones I missed, I'll check later.

Still have to check optional unions like `Optional[Union[str, int]]` and I will change when pickling unions on 3.6 still fails.

## Tests list
A list of use cases to cover with tests, add more here if you think of more.
- [x] Unions with optional:
```
Optional[Union[int, str]]
Union[int,str,None] # I think this is legal to define Optional
```
- [x] Union with containers, Lists, Dicts
Union[str, List[str]]
- [x] Containers with Unions:
List[Union[int, str]]
- [x] Assignment to union fields, make sure it respects the type and that the field is still a UnionNode
- [x] Merging into Union fields, , make sure it respects the type and that the field is still a UnionNode
- [x] As you pointed out, pickling with Union nodes.

Closes #144 .

omry commented:
Cool.
A few surface area points to test/consider:
- Unions with optional:
```
Optional[Union[int, str]]
Union[int,str,None] # I think this is legal to define Optional
```
- Union with containers, Lists, Dicts
Union[str, List[str]]
- Containers with Unions:
List[Union[int, str]]
- Assignment to union fields, make sure it respects the type and that the field is still a UnionNode
- Merging into Union fields, , make sure it respects the type and that the field is still a UnionNode
- As you pointed out, pickling with Union nodes.

omry commented on omegaconf/_utils.py:
```diff
@@ -446,6 +446,31 @@ def get_union_types(ref_type: Optional[Any]) -> List[Any]:
     return element_types
 
 
+def is_valid_value_list_annotation(value: Any, list_type: List[Any]) -> bool:
+    element_type = get_list_element_type(list_type)  # type: ignore
+    valid_value = True
+    if not isinstance(value, list):
+        return False
+    for item in value:
+        if not isinstance(item, element_type):
+            valid_value = False
+            break
+    return valid_value
+
+
+def is_valid_value_dict_annotation(value: Any, dict_type: Dict[Any, Any]) -> bool:
+    key_type, element_type = get_dict_key_value_types(dict_type)
+    valid_value = True
+    if not isinstance(value, dict):
+        return False
+    for key in value:
+        item = value[key]
+        if not isinstance(key, key_type) or not isinstance(item, element_type):
+            valid_value = False
+            break
+    return valid_value
```
These seems fragile:
what about nested objects?
```
List[User] ?
List[Dict[str, int]] ?
```

ListConfig and DictConfig already knows how to check if their input it valid during set, right?

pereman2 replied:
I'm changing it you're right.

omry commented on omegaconf/nodes.py:
```diff
@@ -208,13 +208,23 @@ def validate_and_convert(self, value: Any) -> Optional[Any]:
+                valid_type = is_valid_value_dict_annotation(value, element_type)
+            elif element_type == ListConfig and type_ == list:
                 valid_type = True
             elif element_type == DictConfig and type_ == dict:
```
shouldn't this be "is DictConfig" ?

omry commented on omegaconf/_utils.py:
```diff
@@ -771,3 +746,76 @@ def is_generic_dict(type_: Any) -> bool:
+
+    from .base import Node
+
+    valid_type = False
```
I am not sure about it. seems like you will end up replicating the validation logic in the existing nodes/DictConfig/ListConfig.

For example, IntegerNode accepts "1" as a valid integer.
DictConfig supports both User (class) and User() as a valid input.

omry replied:
Also, what about MISSING (???), what about None?
I suspect you should somehow reuse the existing validation logic here and not recreate it.

pereman2 replied:
Assignment of new values aren't wrapped nodes so I thought it should check against the type annotation.

omry replied:
which would make this illegal?

```
@dataclass
class C:
  x : Union[int, float] = MISSING
```

pereman2 commented:
I've done several things. I've removed validation with other functions since as you pointed out, it doesn't make much sense.
I've done Containers with Union element_types with more or less test. The same goes to Optional and merge with union node.

I'll try to add more test cases and wait for your review because I think you might want to comment a lot.

Still pickle not done.

omry commented:
Re-request review when ready:
![image](https://user-images.githubusercontent.com/376455/99154715-b7f8fd00-2666-11eb-9ed8-2cffef8775cb.png)

pereman2 commented:
Hey! I'll comment every change by the tasks:

1. Union with containers, Lists, Dicts
Union[str, List[str]]

I've removed the function (662772bfbf6603145d5ad647275688436ac40197) to check whether the value is a valid type and changed it so it's checked when wraping the value with each element\_type(until it finds a valid one, if not, it raises error).

2. Containers with Unions: (662772bfbf6603145d5ad647275688436ac40197)
List[Union[int, str]]

In this case it didn't need that much work. I you add a new value to a DictConfig or ListConfig it will wrap with type\_ Union and union will handle the rest.

3. Optional[Union[int, str]] \(19a4012d1093705520234c0e1ef5c1826e039627)

Optional[Union[int ,str]] returns Union[int, str, NoneType] directly so if none is inside Union, UnionNode's optional will be true then, it removes None from Union[int, str] because optional it's checked in Node's validate.

4. Assignment to union fields, make sure it respects the type and that the field is still a UnionNode 

I added test for this one in 9037ea0728e5c3f42dbf26257feea00270b86624 and the tests passed.

5. Merge with UnionNodes (9037ea0728e5c3f42dbf26257feea00270b86624)

After adding tests here I foound that I didn't test structured config assignment to union nodes and I had to add specific validation for it. Moreover since I use wrap node to check if a value is correct I didn't want to wrap Nodes again(it will return the node and not check) so,I used the value of the node to wrap.

6. Pickle with UnionNodes. 

The way I faced this is transforming Union[vt1, vt2] to the list of element\_types and then when loading I use the list to reconstruct the Union type as before. It looks strange but for example: `u = Union[str, int]` and then `u = Union[u, float]` will return `Union[str, int, Float.

I believe there's refactoring to do in `validate_and_convert()` and more test cases maybe but, I think if you review this now it'll be better.

omry reviewed (commented):
Can you check the list items?
![image](https://user-images.githubusercontent.com/376455/99722815-160e5180-2a66-11eb-985c-38ab47d6f1fd.png)

pereman2 commented on tests/test_serialization.py:
```diff
@@ -148,6 +151,35 @@ def test_pickle(obj: Any, ref_type: Any) -> None:
+            assert node._metadata.element_type == ref_type
+        else:
+            assert node._metadata.ref_type == ref_type
+            assert collections.Counter(node.element_types) == collections.Counter(
+                get_union_types(ref_type)
+            )
```
> Also, _bring_type_to_front is implemented in case there's a value that can be converted in more than 1 element_type for example, 1 can be a str and an int therefore, if 1 is an int it should wrap first a IntegerNode instead of StringNode which would be incorrect.

Being type to front modifies the element_types so with counter it checks id the list has the same elements

odelalleau replied:
Just a note that this may be changed (see my comment regarding modifying element_types)

odelalleau replied:
Resolving since the counter is not needed anymore

pereman2 commented on tests/test_nodes.py:
```diff
@@ -560,3 +568,87 @@ def test_allow_objects() -> None:
+            {"foo": 3.14},
+        ),
+        (
+            UnionNode(element_types=[bool, float], is_optional=True),
```
I'll look into it

pereman2 replied:
Done, ref_type stays as Union[NoneType, X], elements_types is without Nonetype 'Union[x]' and is_optional is True if NoneType is inside ref_type.

odelalleau replied:
So actually, I just suggestd to forbid `Union[NoneType, X]` to keep things simple.
@omry do you see a good reason to support it?

omry replied:
Union[X] is exactly the same as Union[NoneType, X]:
```
In [1]: from typing import Union, Optional

In [2]: class X: ...

In [3]: Optional[X]
Out[3]: typing.Union[__main__.X, NoneType]

In [4]: Union[X, type(None)] == Optional[X]
Out[4]: True

In [5]: Union[type(None), X] == Optional[X]
Out[5]: True
```

optional fields has special handling in OmegaConf (because their support predated Union support) but in theory it could be implemented in terms of a Union instead.

odelalleau replied:
> Optional[X] is exactly the same as Union[NoneType, X]:

Yes, that's why I suggested to support only `Optional[X]` and not `Union[NoneType, X]`: there is no need to support both.

Although we could probably support both, it makes things a bit more complex because now we have a `Union` that shouldn't translated into a `UnionNode` (it should instead be an `XNode` with `optional=True`).

Keeping this conversation resolved but feel free to reopen if you disagree.

odelalleau replied:
Actually, unresolving this conversation because I just realized that in Python, `Optional[X]` is actually translated into `Union[NoneType, X]`. Thus I assume that OmegaConf is already catching this kind of Union somewhere to turn it into an `XNode` with `optional=True`. As a result, the error triggered at https://github.com/pereman2/omegaconf/blob/7d6e39ebed85279ef9cae6cd91bcb77748b2df22/omegaconf/_utils.py#L457 should never happen => it probably should be an assert (I don't think there was a test for it, actually).

pereman2 commented:
I added more test cases for Union inside Dict and List

odelalleau commented on tests/structured_conf/test_structured_config.py:
```diff
@@ -1051,3 +1047,100 @@ def test_update_userlist(
+            cfg._get_node("author")._metadata.ref_type
+            == Union[str, List[str], type(None)]
+        )
+        assert cfg._get_node("author")._metadata.optional is True
```
Minor simplification?
```suggestion
        assert cfg._get_node("author")._is_optional()
```

odelalleau commented on tests/test_basic_ops_dict.py:
```diff
@@ -733,3 +735,23 @@ def test_setdefault() -> None:
+def test_set_valid_union_value() -> None:
+    cfg = OmegaConf.structured(DictUnion)
+    cfg.dict = {"a": 4, "b": 3.14}
+    assert isinstance(cfg.dict._get_node("a"), UnionNode)  # type: ignore
```
Could be removed since it is duplicated two lines below?

pereman2 replied:
Idem below.

odelalleau commented on tests/test_basic_ops_dict.py:
```diff
@@ -733,3 +735,23 @@ def test_setdefault() -> None:
+    assert cfg.dict == {"a": 4, "b": 3.14}
+    assert isinstance(cfg.dict._get_node("a"), UnionNode)  # type: ignore
+    assert isinstance(cfg.dict._get_node("b"), UnionNode)  # type: ignore
+    assert cfg.dict.a == 4  # type: ignore
```
I realized when looking at this line that although `cfg.dict` is declared to be of type `Dict`, OmegaConf actually turns it into a `DictConfig`.
@omry is this intended? (I don't think this has anything to do with this PR)

odelalleau commented on tests/test_basic_ops_list.py:
```diff
@@ -598,3 +598,23 @@ def test_getattr() -> None:
+def test_set_valid_union_value() -> None:
+    cfg = OmegaConf.structured(ListUnion)
+    cfg.list = [4, 3.14]
+    assert isinstance(cfg.list._get_node(0), UnionNode)  # type: ignore
```
Could be removed since duplicated two lines below?

pereman2 replied:
Where is this duplicated? This method is in test_basic_ops_dict and test_basic_ops_list. Maybe test_structured_config?

odelalleau replied:
Two lines below there is exactly the same assert, and in-between only a read access to the config so there's no reason for it to change type (unless you actually want to check that executing `cfg.list == [4, 3.14]` doesn't change the node type? If yes why not, it just looked weird to me)

pereman2 replied:
Oh shoot true.

odelalleau commented on tests/test_nodes.py:
```diff
@@ -560,3 +568,87 @@ def test_allow_objects() -> None:
+        ),
+    ],
+)
+def test_valid_value_union_node(node: ValueNode, value: Any, expected: Any) -> None:
```
Since `expected` is always equal to `value`, it seems that the test could be made a bit simpler by having a single argument `value` and setting `expected = deepcopy(value)`.

odelalleau commented on tests/test_nodes.py:
```diff
@@ -560,3 +568,87 @@ def test_allow_objects() -> None:
+        (UnionNode(element_types=[int, str], value="str"), User()),
+        (UnionNode(element_types=[int, ListConfig]), [1, 2]),
+        (UnionNode(element_types=[int, ListConfig]), {"foo": "var"}),
+        (UnionNode(element_types=[int, DictConfig]), {"foo": "var"}),
```
While trying to understand why assigning a dict was not allowed, I came across the following behavior I find surprising.

When initializing an `Union[DictConfig, int]` with an int, the node is a Union as expected: 

```python
@dataclass
class MyConf1:
    x: Union[DictConfig, int] = 0

c1 = OmegaConf.structured(MyConf1)
print(type(c1._get_node("x")))  # ==> UnionNode (as expected)
```

But when initializing it with a `DictConfig`, it is not a Union anymore:

```python
@dataclass
class MyConf2:
    x: Union[DictConfig, int] = field(default_factory=lambda: DictConfig({"a": 0}))

c2 = OmegaConf.structured(MyConf2)
print(type(c2._get_node("x")))  # DictConfig!
```

Is that working as intended?

pereman2 replied:
hmmmm no, I forgot to test DictConfig and ListConfig as types don't change from UnionNode. I'll check it.

pereman2 replied:
Check 45f72c0abfdcce6a495463bab845c65611644efd. I added support for this and tests.

odelalleau commented on tests/test_serialization.py:
```diff
@@ -148,6 +151,35 @@ def test_pickle(obj: Any, ref_type: Any) -> None:
+
+    from omegaconf._utils import get_union_types
+
+    with tempfile.TemporaryFile() as fp:
```
You could use `pickle.loads` and `pickle.dumps` to avoid having to mess with files.

omry replied:
nice, we should probably clean up the pickle serialization tests.

odelalleau reviewed (changes requested):
I'm not really familiar with the structured config code so there's some stuff I'm not too sure about.

pereman2 commented on omegaconf/omegaconf.py:
```diff
@@ -787,6 +790,21 @@ def _node_wrap(
+        element_types = get_union_types(type_)
+        if type(None) in element_types:
+            is_optional = True
+            element_types.remove(type(None))
```
I agree this could be a problem but, I didn't get to it because, I thought no one would type something with Union less than 2 types.

odelalleau replied:
I agree it may not be worth spending time on it => maybe you could just add an explicit check somewhere (either in the `UnionNode` constructor or in `get_union_types`?) that would raise an exception if someone tries to use `Union[type(None), T]`? (saying they should use `Optional[T]` instead)

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +163,126 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+
+    def __getstate__(self) -> Dict[str, Any]:
+        node = copy.copy(self.__dict__)
+        node["_metadata"].ref_type = Union
```
You're right I'm changing it.

odelalleau replied:
Saw your fix, looks good, however I was wondering why you had to override `__{get,set}state__` in the first place, and it seems that all tests pass if I delete them. Could you please check whether they are really needed? (and if yes, add a test that demonstrates it)

pereman2 replied:
Ah, that is because in python 3.6 pickle doesn't like subscripted types(Dict[kt, vt], List[vt], Union[vt1, vt2]...). If you run the test with 3.6 you'll see that it fails without it.

odelalleau replied:
Oh ok thanks, resolving then! :)

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +163,126 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+        else:
+            return self._value() == other  # type: ignore
+
+    def __ne__(self, other: Any) -> bool:
+        return not self.__eq__(other)
```
I added this because LGTM I think.

odelalleau replied:
Oh ok, not sure why it would complain since the super class essentially does this, but anyway, it's not a big deal => resolving.

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +163,126 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+            else:
+                self._raise_invalid_value(value)
+        if value_type in self.element_types:
+            self._bring_type_to_front(value_type)
```
Oh true, I'm changing it to instead of mutate the list to get an ordered one

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +163,126 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+                return value
+            else:
+                self._raise_invalid_value(value)
+        if value_type in self.element_types:
```
> (actually this would also require changing some of OmegaConf's internals, because currently typed nodes like IntegerNode, FloatNode, etc. don't support subclasses => maybe it should be part of another issue)

Yeah that's the thing with these Nodes otherwise it would be ok.

odelalleau replied:
Resolving then, @omry feel free to think about supporting subclasses more generally in typed nodes :)

pereman2 commented on omegaconf/_utils.py:
```diff
@@ -454,6 +439,16 @@ def is_primitive_container(obj: Any) -> bool:
+def get_union_types(ref_type: Optional[Any]) -> List[Any]:
+    args = getattr(ref_type, "__args__", None)
+    element_types: List[Any]
+    if ref_type is not Union and args is not None and args[0] is not Any:
```
I've changed it to return Any in that case, tell me what you think.

odelalleau replied:
> I've changed it to return Any in that case, tell me what you think.

I think an error would be better, otherwise you have to deal with the return value of `get_union_types` being `Any` everywhere else, and this would make the code more complex for no added value. I see no reason to support it (it's more likely to hide problems than to enable more functionalities).
(also, just trying the latest code (720e405) I'm running into a crash when using `x: Union[Any, int] = 1`)

pereman2 replied:
I added an exception for Any as element type. The same goes to `Union[None, vt]` and `Union[vt]`

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +163,126 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+        if isinstance(value, BaseContainer) and is_structured_config(
+            value._metadata.ref_type
+        ):
+            if value._metadata.ref_type in self.element_types:
```
True, this is much better, subclasses would have not worked.

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +163,126 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+        value: Any,
+        key: Any = None,
+        is_optional: bool = True,
+        parent: Optional[Container] = None,
```
You're right, I'll change it.

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +163,126 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+            self._bring_type_to_front(value_type)
+        for union_type in self.element_types:
+            try:
+                print(union_type, value)
```
Yep my bad.

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +163,126 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+        parent: Optional[Container] = None,
+        is_optional: bool = True,
+        ref_type: Any = Union[Any],
+        element_types: List[Any] = None,  # type: ignore
```
`[]`was the default before but, mypy didn't like that because it was mutable and I changed it.

odelalleau replied:
Right, [] is indeed discouraged, I can think of two "clean" options:
1. (The most common solution in such a situation) `self.element_types = element_types if element_types is not None else []`
2. (Make it a mandatory keyword arg, which may be better here because a `Union` with empty element_types makes no sense anyway):
```python
    def __init__(
        self,
        value: Any = None,
        key: Any = None,
        parent: Optional[Container] = None,
        is_optional: bool = True,
        ref_type: Any = Union[Any],
        *,
        element_types: List[Any],
    ):
```

pereman2 replied:
I like the second option.

pereman2 commented on omegaconf/basecontainer.py:
```diff
@@ -110,6 +116,13 @@ def __setstate__(self, d: Dict[str, Any]) -> None:
             key_type = d["_metadata"].key_type
         element_type = d["_metadata"].element_type
+        if isinstance(element_type, list):
+            union = Union[element_type[0]]  # type: ignore
+            for type_ in element_type:
+                union = Union[union, type_]  # type: ignore
+            d["_metadata"].element_type = union
+
+        element_type = d["_metadata"].element_type
```
Oh nice, I didn't know that.

odelalleau replied:
I don't think you actually fixed that one yet, did you?

pereman2 replied:
Done now

pereman2 commented on omegaconf/_utils.py:
```diff
@@ -454,6 +439,16 @@ def is_primitive_container(obj: Any) -> bool:
+    element_types: List[Any]
+    if ref_type is not Union and args is not None and args[0] is not Any:
+        element_types = list(args)
+    assert element_types is not None
```
> Not sure what the if ref_type is not Union test is supposed to catch: if I remove it, all tests still pass

It tested Unions without types(which don't make sense, like List or Dict alone now that I think about it). 

I think this should be better to return Any in case it's not a union like get_element_list. Correct me if I'm wrong.

PS: Thanks for your feedback!

odelalleau replied:
> > Not sure what the if ref_type is not Union test is supposed to catch: if I remove it, all tests still pass
> 
> It tested Unions without types(which don't make sense, like List or Dict alone now that I think about it).
> 
> I think this should be better to return Any in case it's not a union like get_element_list. Correct me if I'm wrong.

Ah ok, personally I think it'd be best to raise an exception if this function is called with anything that's not a `Union[something]`, because I can't think of any use case where this would be intended.
Also I think that `something` should have at least two elements.

omry replied:
What is a union without a type?
like this?

```
@dataclass
class X:
 a: Union
```

This seems like a legitimate Python (at least it runs and mypy --strict does not complain).
not sure what it's supposed to mean though. Unless someone can tell me what this is for, I think we should raise an error for a Union without a list of types.

odelalleau replied:
> I think we should raise an error for a Union without a list of types.

Agreed -- just tested and currently it's accepted.

odelalleau commented on omegaconf/nodes.py:
```diff
@@ -163,6 +164,126 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+            )
+        )
+
+    def _get_element_types_lead_by(self, type_: Any) -> List[Any]:
+        element_types = copy.copy(self.element_types)
+        element_types.insert(0, element_types.pop(element_types.index(type_)))
+        return element_types
```
I realize that performance is not a major concern, but just in case, it's not too hard to avoid some useless copies:
```suggestion
    def _get_element_types_lead_by(self, type_: Any) -> List[Any]:
        pos = self.element_types.index(type_)
        if pos == 0:
            return self.element_types
        element_types = copy.copy(self.element_types)
        element_types[0], element_types[pos] = element_types[pos], element_types[0]
        return element_types
```

odelalleau reviewed (commented):
Thanks for the fixes, added a few more comments based on 720e405

pereman2 commented on omegaconf/omegaconf.py:
```diff
@@ -787,6 +790,21 @@ def _node_wrap(
+    elif _is_union(type_):
+        element_types = get_union_types(type_)
+        if type(None) in element_types:
+            is_optional = True
```
In this case I think we just ignore. We use that flag internally only if i'm correct

odelalleau replied:
Hmm ok, I'm not sure exactly how this flag is set and what are the expectations, but I tried replacing this line with `assert is_optional` and this uncovered a bug in `_resolve_optional()` that doesn't handle Unions properly (this is code that existed prior to this PR). Could you please have a look to get it fixed? (it should be pretty obvious -- it only considers a Union to be optional if it has length 2 and `NoneType` is in second position, which is quite restrictive).
(I'm not actually sure about the best fix -- it depends on whether or not we actually allow `NoneType` in a Union, I haven't looked yet at what you did about that)

odelalleau commented on omegaconf/_utils.py:
```diff
@@ -454,6 +444,27 @@ def is_primitive_container(obj: Any) -> bool:
 
+def get_union_types(ref_type: Optional[Any]) -> Union[List[Any], Any]:
+    args = getattr(ref_type, "__args__", None)
+    element_types: Union[List[Any], Any]
+    error_msg = ""
+    if args is None:
+        error_msg = "{} is not a valid Union type, it should include more than 1 type."
+    elif Any in args:
+        error_msg = (
+            "{} is not a valid Union type, Any is not a valid union element type."
+        )
+    elif type(None) in args and len(args) == 2:
+        error_msg = "{} is not a valid Union type, we recomend using Optional instead."
+    if error_msg:
+        raise UnsupportedValueType(error_msg.format(ref_type))
+
+    if args is not None and Any not in args:
+        element_types = list(args)
+
+    return element_types
```
A few minor tweaks (note the typo fix in `recomend` if you don't copy/paste it exactly):

```suggestion 
    args = getattr(ref_type, "__args__", None)
    error_msg = ""
    if args is None or len(args) < 2:
        error_msg = "{} is not a valid Union type, it should include more than 1 type."
    elif Any in args:
        error_msg = (
            "{} is not a valid Union type, Any is not a valid union element type."
        )
    elif type(None) in args and len(args) == 2:
        error_msg = "{} is not a valid Union type, we recommend using Optional instead."
    if error_msg:
        raise UnsupportedValueType(error_msg.format(ref_type))
```

odelalleau reviewed (commented):
Just a minor comment for now.

I've been spending some time on  45f72c0 today but it's a tricky one, touching OmegaConf's internals I'm not familiar with. Will need a bit more time to wrap my head around it.

odelalleau commented on omegaconf/omegaconf.py:
```diff
@@ -830,10 +848,11 @@ def _maybe_wrap(
     # if already a node, update key and parent and return as is.
     # NOTE: that this mutate the input node!
-    if isinstance(value, Node):
+    if isinstance(value, Node) and ref_type is Any:
```
So, this is the change I'm most concerned about. Until now `_node_wrap()` was never called (I think) with `value` already being a Node, and I'm a bit worried of what might happen if we allow this (especially if the code doesn't explicitly restricts this to unions).

I don't have a good solution yet, I'm still trying to figure out exactly how and when `_maybe_wrap()` is being used thoughout the codebase. But if there was a way to make sure that the changes required to support Union won't have unintended side effects elsewhere, I would be more confident in letting them in.

I may not have time to look more closely at it in the next few days, sorry. @omry feel free to chime in :)

pereman2 replied:
Wrapping of Nodes is causing me trouble.

odelalleau replied:
> Wrapping of Nodes is causing me trouble.

Yeah, honestly, it seems brittle to me, because the rest of the code wasn't written with that in mind. There are a few weird things going on that I can see, and probably more that I can't.

I submitted a PR to your branch in https://github.com/pereman2/omegaconf/pull/1 with a few things I believe make it at least a bit better. For some of my changes I struggled to write a meaningful test as I got deep into some rabbit holes, so eventually I gave up and decided that just "looking better" would be good enough ;)
(in particular, now `_node_wrap()` will return a `UnionNode` as soon as `type_` is a union type, even if called from outside of `UnionNode`)

I doubt we'll manage to get to something 100% satisfying without a deep re-thinking of some OmegaConf internals (or completely changing the approach taken here), so as far as I'm concerned I'm ok with accepting this PR (with my suggested changes), unless @omry thinks otherwise.

omry replied:
IIRC, The reason maybe_wrap exists is exactly to avoid that.
wrap this object if it's not already a node. so generally I think _node_wrap should not be used on nodes (and should probably assert that).

I am going to look at the code after going through the comment and form a more informed opinion.

odelalleau replied:
> I think _node_wrap should not be used on nodes (and should probably assert that).

Just to give some context, currently we need it because otherwise with something like:
```python
 x: Union[int, DictConfig] = field(default_factory=lambda: DictConfig({}))
```
the type of node `x` is `DictConfig` instead of `UnionNode`, and this may break things later when we try to assign other values to `x`.

`UnionNode` is an odd beast because it's a `ValueNode` (not a `Container`) but it wraps another node -- something that didn't exist before.

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +164,132 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+            if is_structured_config(ref_type):
+                value_type = ref_type
+            if any(
+                issubclass(value_type, union_type) for union_type in self.element_types
```
I've been adding tests and it fails as you pointed out. The code you wrote could be useful, I'll try it. 
I'm finishing a sprint this week so sometime this week i expect to have a fix.

pereman2 replied:
So I've been adding tests and encountered wrapping like this: 
```python
node = DictConfig({"foo": "invalid"})
_maybe_wrap(value=node._value(), ref_type=Dict[str, int], is_optional=True, key=None, parent=    None)
```
which fails because "invalid" is AnyNode and validate_set doesn't take it well. I solved this in #443 so this might need to wait that PR.

omry replied:
@odelalleau, ref_type is the annotated type.
for example:
```
@dataclass
class Foo:
 x : BaseClass = SubClass()
```

BaseClass is the ref_type.
Subclass is the type.
ref_type is checked against during assignment, for example.

odelalleau replied:
> So I've been adding tests and encountered wrapping like this:
> 
> ```python
> node = DictConfig({"foo": "invalid"})
> _maybe_wrap(value=node._value(), ref_type=Dict[str, int], is_optional=True, key=None, parent=    None)
> ```
> 
> which fails because "invalid" is AnyNode and validate_set doesn't take it well. I solved this in #443 so this might need to wait that PR.

Not sure I understand, isn't it expected to fail? (since `"invalid"` is not an `int`)

omry replied:
It is expected to fail. and it is fixed in that other PR @pereman2  mentioned.

omry commented on omegaconf/omegaconf.py:
```diff
@@ -760,13 +763,34 @@ def _node_wrap(
+    is_dict = (
+        is_dict_annotation(type_)
+        or type_ is dict
+        or (type_ is Any and type(value) is dict)
```
I think `type_` here should actually be called ref_type.
See how it's used.

pereman2 replied:
I changed it and it fails tests with ref_type

omry commented on omegaconf/omegaconf.py:
```diff
@@ -831,9 +855,10 @@ def _maybe_wrap(
+    # If already a node, update key and parent and return as is.
+    # NOTE: that this mutates the input node!
+    # If `ref_type` is a union we still need to wrap the node into a `UnionNode`.
+    if isinstance(value, Node) and not _is_union(ref_type):
```
I have a concern about this.
can this lead to multiple wrapping in a Union node?

odelalleau replied:
So, right now in the test suite `value` is never a `UnionNode` here at least.

That being said, I do have concerns with unions of unions (ex: `Uniont[int, Union[str, float]]`). I suspect some things may not work properly. However it may not be worth supporting them since we can combine them in a single `Union` (maybe we should catch it and raise an error though, to be safe).

omry replied:
I think Python is doing it automatically. try that in an interpreter.

odelalleau replied:
Oooh nice :) In that case I think it should be ok -- maybe we could add an `assert not isinstance(value, UnionNode)` in the `else:` block to be safe (unless you can see a way to trigger it in a normal use case?)

omry replied:
We can block this until we see an actual use case. Better having someone complain about something raising an exception than have internal structure that goes haywire in certain scenarios.

omry commented on tests/structured_conf/data/attr_classes.py:
```diff
@@ -492,3 +487,39 @@ class UntypedList:
+@attr.s(auto_attribs=True)
+class UnionWithContainer:
+    union_dict: Union[DictConfig, int] = DictConfig({"foo": 1})
+    union_list: Union[ListConfig, int] = ListConfig([1, 2])
```
Add a test for a case for dataclasses as union elements sharing a common base class.
It's weird but people are doing it.

```python
@dataclass
class Base: ...

@dataclass
class Subclass1(Base): ...

@dataclass
class Subclass2(Base): ...

@dataclass
class UnionOfSubclasses:
    foo: Union[Subclass1, Subclass2] 
```

omry reviewed (commented):
I reviewed until the tests.
Looking pretty good so far.
Thanks @pereman2  and thanks @odelalleau  for your help reviewing.

odelalleau commented on omegaconf/nodes.py:
```diff
@@ -163,6 +169,127 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+            )
+        )
+
+    def _get_element_types_lead_by(self, type_: Any) -> List[Any]:
```
It returns the list of union types (`self.element_types`) reordered so that `type_` comes in first position.

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +169,127 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
         return res
 
 
+class UnionNode(ValueNode):
```
I agree. We should do it.

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +169,127 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+        element_types[0], element_types[pos] = element_types[pos], element_types[0]
+        return element_types
+
+    def __eq__(self, other: Any) -> bool:
```
How can it return NotImplemented?

odelalleau replied:
> How can it return NotImplemented?

Scroll down a bit to the comparison operators in https://docs.python.org/3/reference/datamodel.html#basic-customization

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +169,127 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+        return value_node
+
+    def _raise_invalid_value(self, value: Any, value_type: Any) -> None:
+        raise ValidationError(
```
done.

odelalleau replied:
You can get rid of the `return None` after `self._raise_invalid_value(value, type(value))`

odelalleau replied:
Also regarding fstrings, Omry meant something like:
```python
f"Value '{value}' with value_type '{value_type}' is not in '{self.element_types}'"
```

pereman2 replied:
Oh shoot my bad I did it without thinking.

pereman2 commented on omegaconf/nodes.py:
```diff
@@ -163,6 +169,127 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+                pass
+        if not valid_value:
+            self._raise_invalid_value(value, type(value))
+        return value_node
```
done 1b74272cbaba262b57e6bfb21cdaefa8679a4bfc.

pereman2 commented:
Test are failing now because of I added test cases which include the bug solved in #443.

Should the try block be wrapped in a function. There are two instances of this in the same method.

odelalleau commented on tests/structured_conf/test_structured_config.py:
```diff
@@ -1051,3 +1047,130 @@ def test_update_userlist(
+    def test_union_with_subclasses(self, class_type: str) -> None:
+        module: Any = import_module(class_type)
+        cfg = OmegaConf.structured(module.UnionOfSubclasses)
+        cfg.foo = module.Subclass1
+        assert cfg.foo == module.Subclass1()
```
I'm a bit surprised this works, since `cfg.foo` is set to the class and not an instance. What's going on there?

odelalleau commented on omegaconf/nodes.py:
```diff
@@ -163,6 +169,128 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+                    union_type
+                ):
+                    try:
+                        return self._wrap_node(
+                            value=value._value(), ref_type=union_type
```
I'm not sure it's a big deal but I'd rather ask @omry his opinion on this line, because it is changing the original `ref_type` of `value` to `union_type` (not in-place though since we create a new node, this is why it doesn't worry me too much).

omry commented:
FYI:
I am still working on the recursive defaults in Hydra. It's a very large pull request and I am delaying reviewing of other large changes in Hydra and OmegaConf until I am done with it.
I will hopefully get it over with early after the holidays.

pereman2 commented:
Thanks for the info. Anyways, I thought I had more time but I'm full of college work for this vacations.

omry commented:
I bounced the Union support to OmegaConf 2.2, we will resume working on it after 2.1 is released.
Thanks for working on it.

Since many things have changed in 2.1, we will likely start from scratch but the tests will for sure be very useful.
